### DISCUSSION TITLE
Ozone create bucket with EC replication type with s3 client

### DISCUSSION THREAD

**Discussion by Lucas12138** [apache/ozone#6744] [FAQ]
Hi community,

We're using s3 client to create buckets on ozone

Here're some example sudo code:
```
public void createBucket(final String bucket) {
        final CreateBucketRequest createBucketRequest = new CreateBucketRequest(bucket,
                config.region());
        createBucketRequest.setCannedAcl(CannedAccessControlList.Private);
        s3Client.createBucket(createBucketRequest);
    }
```

it looks like it doesn't have any method available to set the replication type and it's default to ratis.

If we want to set it as EC, how can we do that via the s3 client? 

Besides, I saw Hadoop ozone client has the create bucket method as well and it takes 
https://github.com/apache/ozone/blob/master/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/BucketArgs.java
as input. Can we achieve with this? I'm not sure which arg is designated for this. We're using https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-ozone-client/1.1.0

Alternatively, if there's any xml config to enforce that? I checked https://ozone.apache.org/docs/current/feature/erasurecoding.html but it seems it only has cluster level xml config and it's not working because we're using s3 client, which has a default replication config and "Please note, the above configurations will be used only when client does not pass any replication config or bucket does not have any default values."

Thank you!

**Comment by kerneltime** [comment]
> Hi community,
> 
> We're using s3 client to create buckets on ozone
> 
> Here're some example sudo code:
> 
> ```
> public void createBucket(final String bucket) {
>         final CreateBucketRequest createBucketRequest = new CreateBucketRequest(bucket,
>                 config.region());
>         createBucketRequest.setCannedAcl(CannedAccessControlList.Private);
>         s3Client.createBucket(createBucketRequest);
>     }
> ```
> 
> it looks like it doesn't have any method available to set the replication type and it's default to ratis.
> 
> If we want to set it as EC, how can we do that via the s3 client?
> 
> Besides, I saw Hadoop ozone client has the create bucket method as well and it takes https://github.com/apache/ozone/blob/master/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/BucketArgs.java as input. Can we achieve with this? I'm not sure which arg is designated for this. We're using https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-ozone-client/1.1.0
> 
> Alternatively, if there's any xml config to enforce that? I checked https://ozone.apache.org/docs/current/feature/erasurecoding.html but it seems it only has cluster level xml config and it's not working because we're using s3 client, which has a default replication config and "Please note, the above configurations will be used only when client does not pass any replication config or bucket does not have any default values."
> 
> Thank you!

This is a great idea, we could look into implementing custom headers that allow selecting of replication type. We are an Apache project and if you would like we would love it if you offer a patch in S3 Gateway to add this support. We could add `x-ozone-*` namespace to headers for custom Ozone extensions.
